changelog-format.md

```diff
@@ -0,0 +1,62 @@
+- Feature Name: Standarize on the format of changelogs
+- Start Date: 2019-02-15
+- RFC Status: Draft
+- RFC PR: (leave this empty)
+- Salt Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a standard changelog format for software released by the SaltStack project.
+
+# Motivation
+[motivation]: #motivation
+
+The current format for changelog entries at Salt is unstructured and can be difficult to read.
+
+Many users of Salt have complained that they weren't aware of a breaking change which may
+be the result of a poorly structured and inconsistent changelog between releases.
+
+# Design
+[design]: #detailed-design
+
+This RFC standardizes Salt on the 1.0.0 spec of project changelogs outlined at
+https://keepachangelog.com/en/1.0.0/
+
+Specifically, it provides for the following principles:
+
+- Changelogs are for humans, not machines.
+- There should be an entry for every single version.
```

Comment:
Are you proposing that we have just one ChangeLog file with all of the changes for all versions and get rid of each individual release notes file?

Or are you proposing that we include this keepachangelog methodology to each release notes file?

Also is there an example somewhere of someone creating this changelog in an automated fashion from a git log, or would this have to be managed manually? 

also i like this idea overall :) makes a lot more sense for people to consume the information :)

Reply:
Good question. Personally, I think there is a stronger argument for keeping all changes in a single file. I think it's beneficial to have a file where I could, for example, grep for changes to a particular piece of functionality so I can see how it has evolved between versions.

I don't think there is a decent automated approach for changelogs. I've certainly never seen one. IMHO, to do this correctly it must be done manually.

Reply:
> keeping all changes in a single file.

Currently I like how much information is added to our feature releases atleast. For example for the 2019.2.0 release we had a breaking change here: https://docs.saltstack.com/en/develop/topics/releases/2019.2.0.html#non-backward-compatible-change-to-yaml-renderer

And we were able to add a lot of additional information with examples of how to work around this breaking change. After reading the https://keepachangelog.com/en/1.0.0/ website it does not seem like with this format we would add all this information (correct me if i'm wrong) to the changelog. What if we keep a changelog in conjunction with at least our feature releases release notes?  In our changelog we could reference the release notes as well.

I think for our point releases the changelog is perfect as right now as you pointed out it is essentially a nicely formatted git log, with both the PRs and issues.

> I don't think there is a decent automated approach for changelogs

i figured this approach would need to be manual just wanted to confirm. We could try to catch it by adding to the changelog when a bug is fixed, feature added during PR reviews as we try to do right now, but would need to be more adamant about it especially bug fixes. Then the release manager would need to review at every release and make sure everything is added.

Reply:
I don't see anything in the spec that codifies either accepting or rejecting the idea that changes cannot be links (in fact, there are at least links to user profiles, indicating at least links should be a thing)

I think it would be totally reasonable to do something like this:

```
### Changed
- [Non-backward compatible change to YAML render](https://docs.saltstack.com/en/develop/topics/releases/2019.2.0.html#non-backward-compatible-change-to-yaml-renderer)
- Include JID in minion and master logs
- [Windows `runas` no longer requires password](https://docs.saltstack.com/en/path/to/more/info)
```

Reply:
Sounds good :) I'm good with this as it is then.

Reply:
Based on @cachedout's proposal about maintaining changelogs in single file, would we go back and change the existing release notes or would we apply this principle only for new/upcoming ones ?